Preserve sign of thrust_coefficient

clydemcqueen opened the pull request:
# 🦟 Bug fix

Fixes #1399 

## Summary
Preserve the signs of thrust and thrust_coefficient so that ThrustToAngularVec() is the inverse of AngularVelToThrust(). This provides support for clockwise thrusters.

I modified the `ThrusterTest` fixture to also preserve the signs of thrust and thrust_coefficient. I added one test. All tests pass.

`make codecheck` seems to generate a lot of unrelated cppcheck problems, perhaps I have something set up wrong?

`./tools/clang_tidy.sh` seemed to get stuck on `/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/variant` -- I'm not sure what is going on there.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

clydemcqueen commented:
FYI @arjo129 @mabelzhang

clydemcqueen commented:
DCO check failed; I just uploaded my key, so it should succeed next time. Closing & reopening.

arjo129 reviewed (commented):
Thanks for the changes.

To fix the DCO check out the instructions here:
https://github.com/ignitionrobotics/ign-gazebo/pull/1402/checks?check_run_id=5612527595
You need to ` git rebase HEAD~2 --signoff` and the force push `git push --force-with-lease origin sfix_1399`.

Regarding codecheck as long as `make codecheck` returns a zero you can ignore the previous  warnings. 

A few other knits:
* We need to document the default behavior of the propeller being counterclockwise when positive and clockwise when negative in `src/systems/thruster/Thruster.hh`
* We need to verify if this behaviour is consistent when `<use_ang_vel>` is set.

clydemcqueen commented:
Thanks @arjo129. I will add the additional documentation and tests.

clydemcqueen commented:
I think this addresses all of the feedback:
* I added a test to cover the <use_angvel_cmd> case with a negative <thrust_coefficient>.
* I updated the documentation in Thruster.hh to note that a negative <thrust_coefficient> means clockwise rotation.

I also made a few other changes to the documentation in Thruster.hh:
* There were conflicting comments as to the default for <use_angvel_cmd>, I verified that it is false and updated the comment.
* I noted that the parameters <min_thrust_cmd> and <max_thrust_cmd> also apply to angular velocity in the <use_angvel_cmd> case. This is a bit awkward since the units are either N or rad/s. Also, the parameters should probably be named <min_cmd> and <max_cmd>, but deprecating the old parameters and adding new ones seemed like overkill.

I'm happy for other feedback.

arjo129 reviewed (approved):
Thanks for these changes. They look good to me.

chapulina commented on src/systems/thruster/Thruster.hh:
```diff
@@ -53,15 +53,18 @@ namespace systems
-  ///   velocity to actual thrust. [Optional, no units, defaults to 1.0]
+  ///   velocity to actual thrust. A positive coefficient spins the propeller
+  ///   counter-clockwise for a positive thrust force; a negative coefficient
+  ///   spins the propeller clockwise for positive thrust force. [Optional,
```
Just a note that "clockwise" can be confusing unless the reference frame is clear. It's worth clarifying whether this is from the point of view from the vehicle, or looking at it from the outside.

clydemcqueen replied:
Geez, great catch @chapulina

I meant to apply the maritime nomenclature: viewed stern to bow, a clockwise (right-hand) propeller spins clockwise to provide forward (positive) thrust. But I got it flipped. I will fix that, and note the reference frame.

chapulina reviewed (approved):
LGTM, I just have a nitpick about the documentation

clydemcqueen commented:
@chapulina please take another look. ignition_gazebo-ci-pr_any-homebrew-amd64 failed 1 test but for unrelated reasons? Thanks

Rezenders commented:
Hello,

I am trying to build Ignition from source with this patch but I am having problems matching all the packages versions.
If I use `collection-garden.yaml` it gets everything from the main branches, it builds but then I have problems with packages integrating ros and ignition. I guess that is probably expected since the main branches receive updates daily.
So, what I have been trying to do is to figure out which specific commits I should checkout to build garden with this patch. Any suggestions?

clydemcqueen commented:
Hello, @Rezenders 

Please see PR#2 in bluerov2_ignition. It's not complete, but it does include the hashes you are looking for.

Rezenders commented:
Thanks for the reply. I will try it!

jeroenzwan commented:
Hi,
![Missing_files_common](https://user-images.githubusercontent.com/71556561/188656214-eaa9ab4a-7e35-49c2-a9ed-3edf3642c74d.png)

I am trying to also use this version with the hashes you referred to in a previous post. However, when I am compiling it seems to me that there are quite some files missing in the ign-common package. Above is a screenshot of the folders and the content of ImageHeightmap.hh. I am trying to compile using a system with Ubuntu 22.04 and ROS2 Humble.
Below you can see the error that happens when I try to build with `colcon build --merge-install --cmake-args ' -DBUILD_TESTING=OFF' ' -DCMAKE_BUILD_TYPE=Release'`
![Screenshot from 2022-09-06 16-10-19](https://user-images.githubusercontent.com/71556561/188657575-ad7450ab-759c-4cb6-a332-338b32fcae0c.png)
Do you have any suggestions?

mjcarroll commented:
You are missing the `common::graphics` component due to a lack of `freeimage`, `sudo apt install libfreeimage-dev` should fix it.

jeroenzwan commented:
Thanks @mjcarroll for your reply. However, I still have the same error.

jeroenzwan commented:
Update: I was indeed missing libfreeimage-dev, so I went through installing all the dependencies again and found I was missing some others as well. This resolved the issue, thanks for the help!
